test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc

```diff
@@ -2089,9 +2173,301 @@ TEST_P(ProxyProtocolTest, DrainError) {
   write("PROXY TCP4 1.2.3.4 253.253.253.253 65535 1234\r\nmore data");
 
   expectProxyProtoError();
+  EXPECT_EQ(stats_store_.counter("downstream_cx_proxy_proto.v1.error").value(), 1);
 }
 #endif
 
+class ProxyProtocolDisallowedVersionsTest : public ProxyProtocolTest {
+public:
+  virtual envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol
+  createConfig(const std::vector<ProxyProtocolConfig::Version>& disallowed_versions) const {
+    envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config;
+    for (const auto& version : disallowed_versions) {
+      proto_config.mutable_disallowed_versions()->Add(version);
+    }
+    return proto_config;
+  }
+};
+
+// Parameterize the listener socket address version.
+INSTANTIATE_TEST_SUITE_P(IpVersions, ProxyProtocolDisallowedVersionsTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(ProxyProtocolDisallowedVersionsTest, V1DisallowedV2BasicAllowed) {
+  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config =
+      createConfig({ProxyProtocolConfig::V1});
+  connect(true, &proto_config);
+
+  // A well-formed ipv4/tcp message, no extensions
+  constexpr uint8_t buffer[] = {0x0d, 0x0a, 0x0d, 0x0a, 0x00, 0x0d, 0x0a, 0x51, 0x55, 0x49,
+                                0x54, 0x0a, 0x21, 0x11, 0x00, 0x0c, 0x01, 0x02, 0x03, 0x04,
+                                0x00, 0x01, 0x01, 0x02, 0x03, 0x05, 0x00, 0x02, 'm',  'o',
+                                'r',  'e',  ' ',  'd',  'a',  't',  'a'};
+  write(buffer, sizeof(buffer));
+  expectData("more data");
+  disconnect();
+  EXPECT_EQ(stats_store_.counter("downstream_cx_proxy_proto.v2.allowed").value(), 1);
+}
+
+TEST_P(ProxyProtocolDisallowedVersionsTest, V2DisallowedV2BasicRejected) {
+  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config =
+      createConfig({ProxyProtocolConfig::V2});
+  connect(false, &proto_config);
+
+  // A well-formed ipv4/tcp message, no extensions
+  constexpr uint8_t buffer[] = {0x0d, 0x0a, 0x0d, 0x0a, 0x00, 0x0d, 0x0a, 0x51, 0x55, 0x49,
+                                0x54, 0x0a, 0x21, 0x11, 0x00, 0x0c, 0x01, 0x02, 0x03, 0x04,
+                                0x00, 0x01, 0x01, 0x02, 0x03, 0x05, 0x00, 0x02, 'm',  'o',
+                                'r',  'e',  ' ',  'd',  'a',  't',  'a'};
+  write(buffer, sizeof(buffer));
+  expectConnectionError();
+  EXPECT_EQ(stats_store_.counter("downstream_cx_proxy_proto.v2.denied").value(), 1);
+}
+
+TEST_P(ProxyProtocolDisallowedVersionsTest, V1V2DisallowedV2BasicRejected) {
+  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config =
+      createConfig({ProxyProtocolConfig::V1, ProxyProtocolConfig::V2});
```

Comment:
Should disallowing all versions even be an allowed configuration? Is there any use case for this?

Reply:
I don't see any use case. But from an API design perspective, I think we want to keep the `disallowed_versions` as a repeated field in the small possibility of PPv3 in the future (?)

Would you recommend removing the `repeated`? Otherwise I can add an explicit `ProtoValidationException` when both versions are set. This will keep the API open for the future.

Reply:
I'm not sure. @htuch wdyt?

Reply:
I see @markdroth approved API, but happy to hear from @htuch

Reply:
Throwing `ProtoValidationException` when both are configured